{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9fd5b54b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cffi\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "from enum import Enum\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from pathlib import Path\n",
    "from modules.auxiliary import DATA_SEQ_DIR\n",
    "from modules.pytrsomix import TRScalculator, TRSanalyzer\n",
    "import matplotlib.pyplot as plt\n",
    "from multiprocessing import Pool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ce9befdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the directory path containing .fasta files: ./data/klebsiella\n",
      "Enter the value for tmin: 50\n",
      "Enter the value for tmax: 30000\n",
      "Enter the value for mode: 1\n"
     ]
    }
   ],
   "source": [
    "# Prompt user for directory path\n",
    "directory_path = input(\"Enter the directory path containing .fasta files: \")\n",
    "\n",
    "# Get absolute path of the directory\n",
    "directory_path = os.path.abspath(directory_path)\n",
    "\n",
    "# Prompt user for tmin, tmax, and mode\n",
    "tmin = int(input(\"Enter the value for tmin: \"))\n",
    "tmax = int(input(\"Enter the value for tmax: \"))\n",
    "mode = int(input(\"Enter the value for mode: \"))\n",
    "\n",
    "# Get list of .fasta files in the directory\n",
    "fasta_files = [file for file in os.listdir(directory_path) if file.endswith(\".fasta\")]\n",
    "\n",
    "# Initialize a list to store TRScalculator instances\n",
    "trs_calculators = []\n",
    "\n",
    "# Iterate over each .fasta file\n",
    "for fasta_file in fasta_files:\n",
    "    # Get absolute path of the fasta file\n",
    "    fasta_path = os.path.join(directory_path, fasta_file).encode()\n",
    "    \n",
    "    # Define the file path for trs.txt (assuming it's located in the working directory)\n",
    "    trs_file = os.path.abspath(\"trs.txt\").encode()\n",
    "    \n",
    "    # Create TRScalculator instance and calculate TRS\n",
    "    trs_calculator = TRScalculator(sequence=fasta_path, trs=trs_file, tmin=tmin, tmax=tmax, mode=mode)\n",
    "    trs_calculator.calculate()\n",
    "    \n",
    "    # Append the TRScalculator instance to the list\n",
    "    trs_calculators.append(trs_calculator)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0fdb63b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>L-NoClass</th>\n",
       "      <th>L-No</th>\n",
       "      <th>LFS</th>\n",
       "      <th>Len(LFS)</th>\n",
       "      <th>L-POS(LFS)</th>\n",
       "      <th>R-POS(LFS)</th>\n",
       "      <th>R-NoClass</th>\n",
       "      <th>R-No</th>\n",
       "      <th>RFS</th>\n",
       "      <th>Len(RFS)</th>\n",
       "      <th>L-POS(RFS)</th>\n",
       "      <th>R-POS(RFS)</th>\n",
       "      <th>&gt;SEQ</th>\n",
       "      <th>Len(SEQ)</th>\n",
       "      <th>GENOME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>CGGCGGCGG</td>\n",
       "      <td>9</td>\n",
       "      <td>2404</td>\n",
       "      <td>2412</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>CGGCGGCGG</td>\n",
       "      <td>9</td>\n",
       "      <td>3439</td>\n",
       "      <td>3447</td>\n",
       "      <td>CATTGGCCAGTCTCGTTTGACGATGCTGTTGCTGCAGCTTGACCAT...</td>\n",
       "      <td>1026</td>\n",
       "      <td>NC_016845.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>CGGCGGCGG</td>\n",
       "      <td>9</td>\n",
       "      <td>3439</td>\n",
       "      <td>3447</td>\n",
       "      <td>9</td>\n",
       "      <td>26</td>\n",
       "      <td>GCAGCAGCA</td>\n",
       "      <td>9</td>\n",
       "      <td>4374</td>\n",
       "      <td>4382</td>\n",
       "      <td>TATCATTTTCCGCCAGCACCGGATCGAGCTGACCGCTGAGCGTCAT...</td>\n",
       "      <td>926</td>\n",
       "      <td>NC_016845.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>26</td>\n",
       "      <td>GCAGCAGCA</td>\n",
       "      <td>9</td>\n",
       "      <td>4374</td>\n",
       "      <td>4382</td>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>CTGCTGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>4819</td>\n",
       "      <td>4827</td>\n",
       "      <td>GCGTCAGGGTGGTCCCCTGGATCTCCGCCGGCAGTTCATAGTGCGG...</td>\n",
       "      <td>436</td>\n",
       "      <td>NC_016845.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>CTGCTGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>4819</td>\n",
       "      <td>4827</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>GCTGCTGCT</td>\n",
       "      <td>9</td>\n",
       "      <td>6481</td>\n",
       "      <td>6489</td>\n",
       "      <td>CCATTGGGTAAATTCCTCGTCGCTCACCTGCAGCGCGCTGGGCACT...</td>\n",
       "      <td>1653</td>\n",
       "      <td>NC_016845.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>GCTGCTGCT</td>\n",
       "      <td>9</td>\n",
       "      <td>6481</td>\n",
       "      <td>6489</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>GGCGGCGGC</td>\n",
       "      <td>9</td>\n",
       "      <td>6778</td>\n",
       "      <td>6786</td>\n",
       "      <td>GAAGCACCCGGAAGCGATTAAAAACCCGTTCTTCCTGCTGGCGCCT...</td>\n",
       "      <td>288</td>\n",
       "      <td>NC_016845.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32107</th>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>TGGTGGTGG</td>\n",
       "      <td>9</td>\n",
       "      <td>5603843</td>\n",
       "      <td>5603851</td>\n",
       "      <td>14</td>\n",
       "      <td>41</td>\n",
       "      <td>TTCTTCTTC</td>\n",
       "      <td>9</td>\n",
       "      <td>5604969</td>\n",
       "      <td>5604977</td>\n",
       "      <td>TGCCGTCGACCATAAACAGCACCCGATCCGCCTGCTCAATCTCCTG...</td>\n",
       "      <td>1117</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32108</th>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>GCTGCTGCT</td>\n",
       "      <td>9</td>\n",
       "      <td>5605021</td>\n",
       "      <td>5605029</td>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>CTGCTGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>5606507</td>\n",
       "      <td>5606515</td>\n",
       "      <td>GAATAATGGTGACCAGGTTGCTGACGATGTAGTACACCACCAGGCC...</td>\n",
       "      <td>1477</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32109</th>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>CTGCTGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>5606507</td>\n",
       "      <td>5606515</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>CCACCACCA</td>\n",
       "      <td>9</td>\n",
       "      <td>5606927</td>\n",
       "      <td>5606935</td>\n",
       "      <td>GGGCTGCGGATTTTTGTCCTGCTCCCAGGCTTGCCAGATCATGAAA...</td>\n",
       "      <td>411</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32110</th>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>CCACCACCA</td>\n",
       "      <td>9</td>\n",
       "      <td>5606927</td>\n",
       "      <td>5606935</td>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "      <td>ACGACGACG</td>\n",
       "      <td>9</td>\n",
       "      <td>5607243</td>\n",
       "      <td>5607251</td>\n",
       "      <td>CGAAATCCATTGGCGGGAGTTCATGTTGACGCAAACGAAAACTTTC...</td>\n",
       "      <td>307</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32111</th>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "      <td>ACGACGACG</td>\n",
       "      <td>9</td>\n",
       "      <td>5607243</td>\n",
       "      <td>5607251</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>GCGGCGGCG</td>\n",
       "      <td>9</td>\n",
       "      <td>454</td>\n",
       "      <td>462</td>\n",
       "      <td>TGCCAGAACCTGACGACCATTTTTAGTAGCCATACGAGCACGGAAG...</td>\n",
       "      <td>1170</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32112 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       L-NoClass  L-No        LFS  Len(LFS)  L-POS(LFS)  R-POS(LFS)  \\\n",
       "0              2     4  CGGCGGCGG         9        2404        2412   \n",
       "1              2     4  CGGCGGCGG         9        3439        3447   \n",
       "2              9    26  GCAGCAGCA         9        4374        4382   \n",
       "3             10    29  CTGCTGCTG         9        4819        4827   \n",
       "4             10    28  GCTGCTGCT         9        6481        6489   \n",
       "...          ...   ...        ...       ...         ...         ...   \n",
       "32107          4    12  TGGTGGTGG         9     5603843     5603851   \n",
       "32108         10    28  GCTGCTGCT         9     5605021     5605029   \n",
       "32109         10    29  CTGCTGCTG         9     5606507     5606515   \n",
       "32110          3     8  CCACCACCA         9     5606927     5606935   \n",
       "32111          7    19  ACGACGACG         9     5607243     5607251   \n",
       "\n",
       "       R-NoClass  R-No        RFS  Len(RFS)  L-POS(RFS)  R-POS(RFS)  \\\n",
       "0              2     4  CGGCGGCGG         9        3439        3447   \n",
       "1              9    26  GCAGCAGCA         9        4374        4382   \n",
       "2             10    29  CTGCTGCTG         9        4819        4827   \n",
       "3             10    28  GCTGCTGCT         9        6481        6489   \n",
       "4              2     5  GGCGGCGGC         9        6778        6786   \n",
       "...          ...   ...        ...       ...         ...         ...   \n",
       "32107         14    41  TTCTTCTTC         9     5604969     5604977   \n",
       "32108         10    29  CTGCTGCTG         9     5606507     5606515   \n",
       "32109          3     8  CCACCACCA         9     5606927     5606935   \n",
       "32110          7    19  ACGACGACG         9     5607243     5607251   \n",
       "32111          2     6  GCGGCGGCG         9         454         462   \n",
       "\n",
       "                                                    >SEQ  Len(SEQ)  \\\n",
       "0      CATTGGCCAGTCTCGTTTGACGATGCTGTTGCTGCAGCTTGACCAT...      1026   \n",
       "1      TATCATTTTCCGCCAGCACCGGATCGAGCTGACCGCTGAGCGTCAT...       926   \n",
       "2      GCGTCAGGGTGGTCCCCTGGATCTCCGCCGGCAGTTCATAGTGCGG...       436   \n",
       "3      CCATTGGGTAAATTCCTCGTCGCTCACCTGCAGCGCGCTGGGCACT...      1653   \n",
       "4      GAAGCACCCGGAAGCGATTAAAAACCCGTTCTTCCTGCTGGCGCCT...       288   \n",
       "...                                                  ...       ...   \n",
       "32107  TGCCGTCGACCATAAACAGCACCCGATCCGCCTGCTCAATCTCCTG...      1117   \n",
       "32108  GAATAATGGTGACCAGGTTGCTGACGATGTAGTACACCACCAGGCC...      1477   \n",
       "32109  GGGCTGCGGATTTTTGTCCTGCTCCCAGGCTTGCCAGATCATGAAA...       411   \n",
       "32110  CGAAATCCATTGGCGGGAGTTCATGTTGACGCAAACGAAAACTTTC...       307   \n",
       "32111  TGCCAGAACCTGACGACCATTTTTAGTAGCCATACGAGCACGGAAG...      1170   \n",
       "\n",
       "              GENOME  \n",
       "0        NC_016845.1  \n",
       "1        NC_016845.1  \n",
       "2        NC_016845.1  \n",
       "3        NC_016845.1  \n",
       "4        NC_016845.1  \n",
       "...              ...  \n",
       "32107  NZ_CP113789.1  \n",
       "32108  NZ_CP113789.1  \n",
       "32109  NZ_CP113789.1  \n",
       "32110  NZ_CP113789.1  \n",
       "32111  NZ_CP113789.1  \n",
       "\n",
       "[32112 rows x 15 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize an empty list to store the results\n",
    "results_list = []\n",
    "\n",
    "# Iterate over each TRScalculator instance\n",
    "for trs_calculator in trs_calculators:\n",
    "    # Extract the result from the calculator\n",
    "    result = trs_calculator.Result\n",
    "    \n",
    "    # Append the result to the list\n",
    "    results_list.append(result)\n",
    "\n",
    "# Concatenate all results into a single DataFrame\n",
    "combined_results = pd.concat(results_list, ignore_index=True)\n",
    "\n",
    "#Remove \">\" from >SEQ column\n",
    "combined_results[\">SEQ\"] = combined_results[\">SEQ\"].str.replace(\">\", \"\")\n",
    "\n",
    "# Display the combined results\n",
    "combined_results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0443a4d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the name of the file containing results: klebsiella_results.csv\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'combined_results' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-779d3040bf4d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# Save combined_results to CSV file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mcombined_results\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcsv_file_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Results saved to {csv_file_name}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'combined_results' is not defined"
     ]
    }
   ],
   "source": [
    "# Extract directory name from the directory path\n",
    "directory_name = os.path.basename(directory_path)\n",
    "\n",
    "# Define the CSV file name\n",
    "csv_file_name = directory_name + \"_results.csv\"\n",
    "\n",
    "# Save combined_results to CSV file\n",
    "combined_results.to_csv(csv_file_name, index=False)\n",
    "\n",
    "print(f\"Results saved to {csv_file_name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "854ef221",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'directory_name' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-0158077d9ed5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Now we can load \"folder_results.csv\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mfile_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdirectory_name\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"_results.csv\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mcombined_results\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mcombined_results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'directory_name' is not defined"
     ]
    }
   ],
   "source": [
    "# Now we can load \"folder_results.csv\"\n",
    "file_path = \n",
    "combined_results = pd.read_csv(file_path)\n",
    "combined_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fc89e02a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the length of sequence to extract from the start: 50\n",
      "Enter the length of sequence to extract from the end: 50\n"
     ]
    }
   ],
   "source": [
    "def extract_sequences(combined_results, l_chars, r_chars):\n",
    "    # Create new columns SEQ_L and SEQ_R by slicing the sequence column\n",
    "    combined_results['SEQ_L'] = combined_results['>SEQ'].str.slice(0, l_chars)\n",
    "    combined_results['SEQ_R'] = combined_results['>SEQ'].str.slice(-r_chars)    \n",
    "    return combined_results\n",
    "\n",
    "l_chars = int(input(\"Enter the length of sequence to extract from the start: \"))\n",
    "r_chars = int(input(\"Enter the length of sequence to extract from the end: \"))\n",
    "\n",
    "combined_results = extract_sequences(combined_results, l_chars, r_chars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9b23e17c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>L-NoClass</th>\n",
       "      <th>L-No</th>\n",
       "      <th>LFS</th>\n",
       "      <th>Len(LFS)</th>\n",
       "      <th>L-POS(LFS)</th>\n",
       "      <th>R-POS(LFS)</th>\n",
       "      <th>R-NoClass</th>\n",
       "      <th>R-No</th>\n",
       "      <th>RFS</th>\n",
       "      <th>Len(RFS)</th>\n",
       "      <th>L-POS(RFS)</th>\n",
       "      <th>R-POS(RFS)</th>\n",
       "      <th>&gt;SEQ</th>\n",
       "      <th>Len(SEQ)</th>\n",
       "      <th>GENOME</th>\n",
       "      <th>SEQ_L</th>\n",
       "      <th>SEQ_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>CGGCGGCGG</td>\n",
       "      <td>9</td>\n",
       "      <td>2404</td>\n",
       "      <td>2412</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>CGGCGGCGG</td>\n",
       "      <td>9</td>\n",
       "      <td>3439</td>\n",
       "      <td>3447</td>\n",
       "      <td>CATTGGCCAGTCTCGTTTGACGATGCTGTTGCTGCAGCTTGACCAT...</td>\n",
       "      <td>1026</td>\n",
       "      <td>NC_016845.1</td>\n",
       "      <td>CATTGGCCAGTCTCGTTTGACGATGCTGTTGCTGCAGCTTGACCAT...</td>\n",
       "      <td>CTGGTAGTCGCCTCGTTTGAGTTCGCCGGCGCTCATATCCCACAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>CGGCGGCGG</td>\n",
       "      <td>9</td>\n",
       "      <td>3439</td>\n",
       "      <td>3447</td>\n",
       "      <td>9</td>\n",
       "      <td>26</td>\n",
       "      <td>GCAGCAGCA</td>\n",
       "      <td>9</td>\n",
       "      <td>4374</td>\n",
       "      <td>4382</td>\n",
       "      <td>TATCATTTTCCGCCAGCACCGGATCGAGCTGACCGCTGAGCGTCAT...</td>\n",
       "      <td>926</td>\n",
       "      <td>NC_016845.1</td>\n",
       "      <td>TATCATTTTCCGCCAGCACCGGATCGAGCTGACCGCTGAGCGTCAT...</td>\n",
       "      <td>TTCAAAAGTAATGTGGATCACTTCCATATCGTGAAGTTTGAGCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>26</td>\n",
       "      <td>GCAGCAGCA</td>\n",
       "      <td>9</td>\n",
       "      <td>4374</td>\n",
       "      <td>4382</td>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>CTGCTGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>4819</td>\n",
       "      <td>4827</td>\n",
       "      <td>GCGTCAGGGTGGTCCCCTGGATCTCCGCCGGCAGTTCATAGTGCGG...</td>\n",
       "      <td>436</td>\n",
       "      <td>NC_016845.1</td>\n",
       "      <td>GCGTCAGGGTGGTCCCCTGGATCTCCGCCGGCAGTTCATAGTGCGG...</td>\n",
       "      <td>AGCTGGAATATCAGCTCAAACACCGCGTCCGGCAGTTTAATATTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>CTGCTGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>4819</td>\n",
       "      <td>4827</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>GCTGCTGCT</td>\n",
       "      <td>9</td>\n",
       "      <td>6481</td>\n",
       "      <td>6489</td>\n",
       "      <td>CCATTGGGTAAATTCCTCGTCGCTCACCTGCAGCGCGCTGGGCACT...</td>\n",
       "      <td>1653</td>\n",
       "      <td>NC_016845.1</td>\n",
       "      <td>CCATTGGGTAAATTCCTCGTCGCTCACCTGCAGCGCGCTGGGCACT...</td>\n",
       "      <td>TTCTCCGTCGTGCTGCCCTCGTTGGTGCTGAACTATTTTGGCCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>GCTGCTGCT</td>\n",
       "      <td>9</td>\n",
       "      <td>6481</td>\n",
       "      <td>6489</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>GGCGGCGGC</td>\n",
       "      <td>9</td>\n",
       "      <td>6778</td>\n",
       "      <td>6786</td>\n",
       "      <td>GAAGCACCCGGAAGCGATTAAAAACCCGTTCTTCCTGCTGGCGCCT...</td>\n",
       "      <td>288</td>\n",
       "      <td>NC_016845.1</td>\n",
       "      <td>GAAGCACCCGGAAGCGATTAAAAACCCGTTCTTCCTGCTGGCGCCT...</td>\n",
       "      <td>TACGTTTCTGTGGTCATCGTTATCGTCAGCTTTGAGCATTCGTCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32107</th>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>TGGTGGTGG</td>\n",
       "      <td>9</td>\n",
       "      <td>5603843</td>\n",
       "      <td>5603851</td>\n",
       "      <td>14</td>\n",
       "      <td>41</td>\n",
       "      <td>TTCTTCTTC</td>\n",
       "      <td>9</td>\n",
       "      <td>5604969</td>\n",
       "      <td>5604977</td>\n",
       "      <td>TGCCGTCGACCATAAACAGCACCCGATCCGCCTGCTCAATCTCCTG...</td>\n",
       "      <td>1117</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "      <td>TGCCGTCGACCATAAACAGCACCCGATCCGCCTGCTCAATCTCCTG...</td>\n",
       "      <td>ACTAATGAAGCCAACGCATCTGCGGCTTGAAGAATGAAGAGTATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32108</th>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>GCTGCTGCT</td>\n",
       "      <td>9</td>\n",
       "      <td>5605021</td>\n",
       "      <td>5605029</td>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>CTGCTGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>5606507</td>\n",
       "      <td>5606515</td>\n",
       "      <td>GAATAATGGTGACCAGGTTGCTGACGATGTAGTACACCACCAGGCC...</td>\n",
       "      <td>1477</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "      <td>GAATAATGGTGACCAGGTTGCTGACGATGTAGTACACCACCAGGCC...</td>\n",
       "      <td>ACGCCCTGGTCGGCGGCGCTACCCGCTGCTGTGGTCGTAGTCTGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32109</th>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>CTGCTGCTG</td>\n",
       "      <td>9</td>\n",
       "      <td>5606507</td>\n",
       "      <td>5606515</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>CCACCACCA</td>\n",
       "      <td>9</td>\n",
       "      <td>5606927</td>\n",
       "      <td>5606935</td>\n",
       "      <td>GGGCTGCGGATTTTTGTCCTGCTCCCAGGCTTGCCAGATCATGAAA...</td>\n",
       "      <td>411</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "      <td>GGGCTGCGGATTTTTGTCCTGCTCCCAGGCTTGCCAGATCATGAAA...</td>\n",
       "      <td>CAACGCTTCCGAGAGAGCACGGTTATCGAGGTCGGCAACCCCTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32110</th>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>CCACCACCA</td>\n",
       "      <td>9</td>\n",
       "      <td>5606927</td>\n",
       "      <td>5606935</td>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "      <td>ACGACGACG</td>\n",
       "      <td>9</td>\n",
       "      <td>5607243</td>\n",
       "      <td>5607251</td>\n",
       "      <td>CGAAATCCATTGGCGGGAGTTCATGTTGACGCAAACGAAAACTTTC...</td>\n",
       "      <td>307</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "      <td>CGAAATCCATTGGCGGGAGTTCATGTTGACGCAAACGAAAACTTTC...</td>\n",
       "      <td>CAGGGTTAGCTTTATTACTTGGAAACGGTCAGACGAGCGCGGCCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32111</th>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "      <td>ACGACGACG</td>\n",
       "      <td>9</td>\n",
       "      <td>5607243</td>\n",
       "      <td>5607251</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>GCGGCGGCG</td>\n",
       "      <td>9</td>\n",
       "      <td>454</td>\n",
       "      <td>462</td>\n",
       "      <td>TGCCAGAACCTGACGACCATTTTTAGTAGCCATACGAGCACGGAAG...</td>\n",
       "      <td>1170</td>\n",
       "      <td>NZ_CP113789.1</td>\n",
       "      <td>TGCCAGAACCTGACGACCATTTTTAGTAGCCATACGAGCACGGAAG...</td>\n",
       "      <td>AACACACGTTTGATAACTTCGTCGAAGGTAAATCTAACCAGCTGGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32112 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       L-NoClass  L-No        LFS  Len(LFS)  L-POS(LFS)  R-POS(LFS)  \\\n",
       "0              2     4  CGGCGGCGG         9        2404        2412   \n",
       "1              2     4  CGGCGGCGG         9        3439        3447   \n",
       "2              9    26  GCAGCAGCA         9        4374        4382   \n",
       "3             10    29  CTGCTGCTG         9        4819        4827   \n",
       "4             10    28  GCTGCTGCT         9        6481        6489   \n",
       "...          ...   ...        ...       ...         ...         ...   \n",
       "32107          4    12  TGGTGGTGG         9     5603843     5603851   \n",
       "32108         10    28  GCTGCTGCT         9     5605021     5605029   \n",
       "32109         10    29  CTGCTGCTG         9     5606507     5606515   \n",
       "32110          3     8  CCACCACCA         9     5606927     5606935   \n",
       "32111          7    19  ACGACGACG         9     5607243     5607251   \n",
       "\n",
       "       R-NoClass  R-No        RFS  Len(RFS)  L-POS(RFS)  R-POS(RFS)  \\\n",
       "0              2     4  CGGCGGCGG         9        3439        3447   \n",
       "1              9    26  GCAGCAGCA         9        4374        4382   \n",
       "2             10    29  CTGCTGCTG         9        4819        4827   \n",
       "3             10    28  GCTGCTGCT         9        6481        6489   \n",
       "4              2     5  GGCGGCGGC         9        6778        6786   \n",
       "...          ...   ...        ...       ...         ...         ...   \n",
       "32107         14    41  TTCTTCTTC         9     5604969     5604977   \n",
       "32108         10    29  CTGCTGCTG         9     5606507     5606515   \n",
       "32109          3     8  CCACCACCA         9     5606927     5606935   \n",
       "32110          7    19  ACGACGACG         9     5607243     5607251   \n",
       "32111          2     6  GCGGCGGCG         9         454         462   \n",
       "\n",
       "                                                    >SEQ  Len(SEQ)  \\\n",
       "0      CATTGGCCAGTCTCGTTTGACGATGCTGTTGCTGCAGCTTGACCAT...      1026   \n",
       "1      TATCATTTTCCGCCAGCACCGGATCGAGCTGACCGCTGAGCGTCAT...       926   \n",
       "2      GCGTCAGGGTGGTCCCCTGGATCTCCGCCGGCAGTTCATAGTGCGG...       436   \n",
       "3      CCATTGGGTAAATTCCTCGTCGCTCACCTGCAGCGCGCTGGGCACT...      1653   \n",
       "4      GAAGCACCCGGAAGCGATTAAAAACCCGTTCTTCCTGCTGGCGCCT...       288   \n",
       "...                                                  ...       ...   \n",
       "32107  TGCCGTCGACCATAAACAGCACCCGATCCGCCTGCTCAATCTCCTG...      1117   \n",
       "32108  GAATAATGGTGACCAGGTTGCTGACGATGTAGTACACCACCAGGCC...      1477   \n",
       "32109  GGGCTGCGGATTTTTGTCCTGCTCCCAGGCTTGCCAGATCATGAAA...       411   \n",
       "32110  CGAAATCCATTGGCGGGAGTTCATGTTGACGCAAACGAAAACTTTC...       307   \n",
       "32111  TGCCAGAACCTGACGACCATTTTTAGTAGCCATACGAGCACGGAAG...      1170   \n",
       "\n",
       "              GENOME                                              SEQ_L  \\\n",
       "0        NC_016845.1  CATTGGCCAGTCTCGTTTGACGATGCTGTTGCTGCAGCTTGACCAT...   \n",
       "1        NC_016845.1  TATCATTTTCCGCCAGCACCGGATCGAGCTGACCGCTGAGCGTCAT...   \n",
       "2        NC_016845.1  GCGTCAGGGTGGTCCCCTGGATCTCCGCCGGCAGTTCATAGTGCGG...   \n",
       "3        NC_016845.1  CCATTGGGTAAATTCCTCGTCGCTCACCTGCAGCGCGCTGGGCACT...   \n",
       "4        NC_016845.1  GAAGCACCCGGAAGCGATTAAAAACCCGTTCTTCCTGCTGGCGCCT...   \n",
       "...              ...                                                ...   \n",
       "32107  NZ_CP113789.1  TGCCGTCGACCATAAACAGCACCCGATCCGCCTGCTCAATCTCCTG...   \n",
       "32108  NZ_CP113789.1  GAATAATGGTGACCAGGTTGCTGACGATGTAGTACACCACCAGGCC...   \n",
       "32109  NZ_CP113789.1  GGGCTGCGGATTTTTGTCCTGCTCCCAGGCTTGCCAGATCATGAAA...   \n",
       "32110  NZ_CP113789.1  CGAAATCCATTGGCGGGAGTTCATGTTGACGCAAACGAAAACTTTC...   \n",
       "32111  NZ_CP113789.1  TGCCAGAACCTGACGACCATTTTTAGTAGCCATACGAGCACGGAAG...   \n",
       "\n",
       "                                                   SEQ_R  \n",
       "0      CTGGTAGTCGCCTCGTTTGAGTTCGCCGGCGCTCATATCCCACAGG...  \n",
       "1      TTCAAAAGTAATGTGGATCACTTCCATATCGTGAAGTTTGAGCGGC...  \n",
       "2      AGCTGGAATATCAGCTCAAACACCGCGTCCGGCAGTTTAATATTGC...  \n",
       "3      TTCTCCGTCGTGCTGCCCTCGTTGGTGCTGAACTATTTTGGCCAGG...  \n",
       "4      TACGTTTCTGTGGTCATCGTTATCGTCAGCTTTGAGCATTCGTCTA...  \n",
       "...                                                  ...  \n",
       "32107  ACTAATGAAGCCAACGCATCTGCGGCTTGAAGAATGAAGAGTATCA...  \n",
       "32108  ACGCCCTGGTCGGCGGCGCTACCCGCTGCTGTGGTCGTAGTCTGCG...  \n",
       "32109  CAACGCTTCCGAGAGAGCACGGTTATCGAGGTCGGCAACCCCTCTT...  \n",
       "32110  CAGGGTTAGCTTTATTACTTGGAAACGGTCAGACGAGCGCGGCCTT...  \n",
       "32111  AACACACGTTTGATAACTTCGTCGAAGGTAAATCTAACCAGCTGGC...  \n",
       "\n",
       "[32112 rows x 17 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b425ed6b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2eb85307",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter your email address: hsalamaga@ibb.waw.pl\n"
     ]
    }
   ],
   "source": [
    "# Prompt the user to input their email address\n",
    "user_email = input(\"Enter your email address: \")\n",
    "\n",
    "# Set the user's email address (required by NCBI)\n",
    "Entrez.email = user_email\n",
    "\n",
    "# Function to fetch organism names from NCBI based on NCBI IDs\n",
    "def fetch_organism_names(ncbi_ids):\n",
    "    # Initialize a dictionary to store the organism names\n",
    "    organism_map = {}\n",
    "    # Iterate over the NCBI IDs\n",
    "    for ncbi_id in ncbi_ids:\n",
    "        # Fetch the record from NCBI\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=ncbi_id, rettype=\"gb\", retmode=\"text\")\n",
    "        record = handle.read()\n",
    "        # Extract the organism information from the record\n",
    "        for line in record.splitlines():\n",
    "            if line.startswith(\"  ORGANISM\"):\n",
    "                organism = line.split(\"ORGANISM\")[1].strip()\n",
    "                # Replace spaces with underscores\n",
    "                organism = organism.replace(\" \", \"_\")\n",
    "                organism_map[ncbi_id] = organism\n",
    "                break  # Once organism information is found, break the loop\n",
    "    return organism_map\n",
    "\n",
    "# Read the DataFrame with the combined_results\n",
    "ncbi_ids = combined_results[\"GENOME\"].unique().tolist()\n",
    "\n",
    "# Fetch organism names from NCBI\n",
    "organism_map = fetch_organism_names(ncbi_ids)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ad068cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map NCBI IDs to taxonomic names\n",
    "combined_results['Taxonomic_Name'] = combined_results['GENOME'].map(organism_map)\n",
    "\n",
    "# Identify unmatched genomes\n",
    "unmatched_genomes = combined_results[combined_results['Taxonomic_Name'].isnull()]['GENOME'].unique()\n",
    "\n",
    "if len(unmatched_genomes) > 0:\n",
    "    print(f\"Warning: Some genome IDs could not be matched with taxonomic names: {unmatched_genomes}\")\n",
    "\n",
    "# Extract sequences and create sequence IDs\n",
    "combined_results['L_id'] = combined_results['Taxonomic_Name'] + '_L' + combined_results['L-No'].astype(str)\n",
    "combined_results['R_id'] = combined_results['Taxonomic_Name'] + '_R' + combined_results['R-No'].astype(str)\n",
    "\n",
    "# Select relevant columns\n",
    "sequences_df = combined_results[['SEQ_L', 'SEQ_R', 'L_id', 'R_id']]\n",
    "\n",
    "# Save left sequences to FASTA file\n",
    "with open('sequences_L.fasta', 'w') as fasta_file:\n",
    "    for _, row in sequences_df.iterrows():\n",
    "        fasta_file.write(f'>{row[\"L_id\"]}\\n')\n",
    "        fasta_file.write(f'{row[\"SEQ_L\"]}\\n')\n",
    "\n",
    "# Save right sequences to FASTA file\n",
    "with open('sequences_R.fasta', 'w') as fasta_file:\n",
    "    for _, row in sequences_df.iterrows():\n",
    "        fasta_file.write(f'>{row[\"R_id\"]}\\n')\n",
    "        fasta_file.write(f'{row[\"SEQ_R\"]}\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "993692f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#There is possibility of several sequences having the same name as such we need to make sure that all ids are unique\n",
    "def rename_sequences(input_file, output_file):\n",
    "    \"\"\"\n",
    "    Renames sequences in a FASTA file.\n",
    "\n",
    "    Parameters:\n",
    "        input_file (str): Path to the input FASTA file.\n",
    "        output_file (str): Path to the output FASTA file with renamed sequences.\n",
    "    \"\"\"\n",
    "    # Open input and output FASTA files\n",
    "    with open(input_file, \"r\") as input_handle, open(output_file, \"w\") as output_handle:\n",
    "        # Read sequences from input FASTA file and write to output with new names\n",
    "        for i, record in enumerate(SeqIO.parse(input_handle, \"fasta\"), start=1):\n",
    "            # Extract the original ID\n",
    "            original_id = record.id\n",
    "            # Rename the sequence using a combination of original ID and sequence number\n",
    "            new_id = f\"{original_id}_{i}\"\n",
    "            record.id = new_id\n",
    "            record.description = \"\"\n",
    "            # Write the updated record to the output file\n",
    "            SeqIO.write(record, output_handle, \"fasta\")\n",
    "\n",
    "# Call the function to rename sequences\n",
    "rename_sequences(\"sequences_R.fasta\", \"sequences_R_unique.fasta\")\n",
    "rename_sequences(\"sequences_L.fasta\",\"sequences_L_unique.fasta\")\n",
    "# This function incorporates index into sequence name making it easy to find the original sequence\n",
    "# Remove original .fasta files\n",
    "os.remove(\"sequences_R.fasta\")\n",
    "os.remove(\"sequences_L.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4fac452a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "def run_cdhit(cdhit_path, input_file, output_file, c=0.9, n=5, d=0, m=\"t\",g=0,G=1,sc=0):\n",
    "    \"\"\"\n",
    "    Run CD-HIT program from Python.\n",
    "\n",
    "    Parameters:\n",
    "        cdhit_path (str): Path to the CD-HIT executable.\n",
    "        input_file (str): Path to the input FASTA file.\n",
    "        output_file (str): Path to the output file.\n",
    "        c (float): Sequence identity threshold. Default is 0.9.\n",
    "        n (int): Word length, default is 5.\n",
    "        d (int): Bandwidth of alignment, default is 0.\n",
    "        m (str): Memory limit, default is \"t\" for unlimited.\n",
    "        g (int): def 0 \n",
    "        G (int): def 0\n",
    "        sc (int): def 0\n",
    "    \"\"\"\n",
    "    # Construct the CD-HIT command\n",
    "    cmd = [cdhit_path, \"-i\", input_file, \"-o\", output_file, \"-c\", str(c), \"-n\", str(n), \"-d\", str(d), \"-M\", m,\n",
    "          \"-g\",str(g), \"-G\", str(G),\"-sc\",str(sc)]\n",
    "\n",
    "    # Run CD-HIT command as a subprocess\n",
    "    subprocess.run(cmd)\n",
    "\n",
    "# Example usage:\n",
    "run_cdhit(\"/home/hubert/cdhit-master/cd-hit-est\",\"sequences_L_unique.fasta\",\"sequences_L_unique_cdhit\",\n",
    "          c=1, n=5, d=0, m=\"t\",g=1,sc=1)\n",
    "run_cdhit(\"/home/hubert/cdhit-master/cd-hit-est\",\"sequences_R_unique.fasta\",\"sequences_R_unique_cdhit\",\n",
    "          c=1, n=5, d=0, m=\"t\",g=1,sc=1)\n",
    "os.remove(\"sequences_L_unique_cdhit\")\n",
    "os.remove(\"sequences_R_unique_cdhit\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a2be51b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_sequence_ids(input_file, output_file):\n",
    "    \"\"\"\n",
    "    Extract sequence IDs from clusters with more than 2 sequences and save them to a file.\n",
    "\n",
    "    Parameters:\n",
    "        input_file (str): Path to the input cluster data file.\n",
    "        output_file (str): Path to the output file to save sequence IDs.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    with open(input_file, \"r\") as f, open(output_file, \"w\") as outfile:\n",
    "        current_cluster = []\n",
    "        for line in f:\n",
    "            if line.startswith(\">Cluster\"):\n",
    "                if len(current_cluster) >= 2:\n",
    "                    for seq in current_cluster:\n",
    "                        outfile.write(seq)\n",
    "                current_cluster = []  # Clear current cluster\n",
    "            elif line.strip():  # Proceed if the line is not empty\n",
    "                current_cluster.append(line)  # Add sequence to current cluster\n",
    "        if len(current_cluster) >= 2:\n",
    "            for seq in current_cluster:\n",
    "                outfile.write(seq)\n",
    "# Example usage\n",
    "input_file = \"sequences_L_unique_cdhit.clstr\"\n",
    "output_file = \"L_clusters.txt\"\n",
    "extract_sequence_ids(input_file, output_file)\n",
    "\n",
    "input_file = \"sequences_R_unique_cdhit.clstr\"\n",
    "output_file =\"R_clusters.txt\"\n",
    "extract_sequence_ids(input_file,output_file)\n",
    "\n",
    "\n",
    "def clean_sequence_ids(input_file):\n",
    "    \"\"\"\n",
    "    Clean sequence IDs file by removing prefixes and suffixes.\n",
    "\n",
    "    Parameters:\n",
    "        input_file (str): Path to the input file with sequence IDs.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    with open(input_file, \"r\") as infile:\n",
    "        lines = infile.readlines()\n",
    "\n",
    "    with open(input_file, \"w\") as outfile:\n",
    "        for line in lines:\n",
    "            sequence_id = line.split(\">\")[1].split(\"...\")[0].strip()\n",
    "            outfile.write(\">\" + sequence_id + \"\\n\")\n",
    "\n",
    "# Example usage\n",
    "input_file = \"L_clusters.txt\"\n",
    "clean_sequence_ids(input_file)\n",
    "\n",
    "input_file = \"R_clusters.txt\"\n",
    "clean_sequence_ids(input_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "47cdf0e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered file created successfully!\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Remove entries in L/R_clusters from the .fasta file \n",
    "\"\"\"\n",
    "def read_fasta_ids(filename):\n",
    "    fasta_ids = set()\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('>'):\n",
    "                fasta_id = line.strip()[1:]\n",
    "                fasta_ids.add(fasta_id)\n",
    "    return fasta_ids\n",
    "\n",
    "def filter_fasta_file(input_file, output_file, fasta_ids_to_remove):\n",
    "    with open(input_file, 'r') as f_in, open(output_file, 'w') as f_out:\n",
    "        remove_entry = False\n",
    "        for line in f_in:\n",
    "            if line.startswith('>'):\n",
    "                current_fasta_id = line.strip()[1:]\n",
    "                if current_fasta_id in fasta_ids_to_remove:\n",
    "                    remove_entry = True\n",
    "                else:\n",
    "                    remove_entry = False\n",
    "            if not remove_entry:\n",
    "                f_out.write(line)\n",
    "\n",
    "# Read fasta IDs from the L_clusters file\n",
    "fasta_ids_to_remove = read_fasta_ids(\"L_clusters.txt\")\n",
    "\n",
    "# Filter the sequences_L_unique.fasta file\n",
    "filter_fasta_file(\"sequences_L_unique.fasta\", \"filtered_sequences_L_unique.fasta\", fasta_ids_to_remove)\n",
    "\n",
    "print(\"Filtered file created successfully!\")\n",
    "\n",
    "fasta_ids_to_remove = read_fasta_ids(\"R_clusters.txt\")\n",
    "\n",
    "filter_fasta_file(\"sequences_R_unique.fasta\", \"filtered_sequences_R_unique.fasta\", fasta_ids_to_remove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3c5c29c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Left cluster sequences file created successfully!\n",
      "Right cluster sequences file created successfully!\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Extract only ids matching those in L/R Cluster files they might come in handy \n",
    "and will be helpful to verify whether or not our way of thinking is valid\n",
    "\"\"\"\n",
    "def filter_fasta_file_clusters(input_file, output_file, fasta_ids_to_include):\n",
    "    with open(input_file, 'r') as f_in, open(output_file, 'w') as f_out:\n",
    "        include_entry = False\n",
    "        for line in f_in:\n",
    "            if line.startswith('>'):\n",
    "                current_fasta_id = line.strip()[1:]\n",
    "                if current_fasta_id in fasta_ids_to_include:\n",
    "                    include_entry = True\n",
    "                    f_out.write(line)\n",
    "                else:\n",
    "                    include_entry = False\n",
    "            elif include_entry:\n",
    "                f_out.write(line)\n",
    "\n",
    "# Read fasta IDs from the L_clusters file\n",
    "fasta_ids_to_include_L = read_fasta_ids(\"L_clusters.txt\")\n",
    "\n",
    "# Filter the sequences_L_unique.fasta file\n",
    "filter_fasta_file_clusters(\"sequences_L_unique.fasta\", \"sequences_L_in_clusters.fasta\", fasta_ids_to_include_L)\n",
    "\n",
    "print(\"Left cluster sequences file created successfully!\")\n",
    "\n",
    "# Read fasta IDs from the R_clusters file\n",
    "fasta_ids_to_include_R = read_fasta_ids(\"R_clusters.txt\")\n",
    "\n",
    "# Filter the sequences_R_unique.fasta file\n",
    "filter_fasta_file_clusters(\"sequences_R_unique.fasta\", \"sequences_R_in_clusters.fasta\", fasta_ids_to_include_R)\n",
    "\n",
    "print(\"Right cluster sequences file created successfully!\")\n",
    "#for some reason the order of sequences is different but it works nonetheless"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11984d2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:TRS] *",
   "language": "python",
   "name": "conda-env-TRS-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
